src/xian/xian_abci.py

```diff
@@ -217,57 +204,39 @@ def deliver_tx(self, tx_raw) -> ResponseDeliverTx:
-                code=result["tx_result"]["status"],
-                data=encode_str(parsed_tx_result),
-                gas_used=result["stamp_rewards_amount"],
-            )
-        except Exception as err:
-            logger.error(f"DELIVER TX ERROR: {err}")
-            return ResponseDeliverTx(code=ErrorCode)
-
-    def end_block(self, req: RequestEndBlock) -> ResponseEndBlock:
-        """
-        Called at the end of processing the current block. If this is a stateful application
-        you can use the height from the request to record the last_block_height
-        """
-
-        return ResponseEndBlock(validator_updates=self.validator_handler.build_validator_updates())
-
-    def commit(self) -> ResponseCommit:
-        """
-        Called after ``end_block``.  This should return a compact ``fingerprint``
-        of the current state of the application. This is usually the root hash
-        of a merkletree.  The returned data is used as part of the consensus process.
-
-        Save all cached state from the block to filesystem DB
-        """
-
-        # a hash of the previous block's app_hash + each of the tx hashes from this block.
-        fingerprint_hash = hash_list(self.fingerprint_hashes)
+            tx_results.append(ExecTxResult(code=result["tx_result"]["status"],data=encode_str(parsed_tx_result),gas_used=result["stamp_rewards_amount"]))
 
         if self.static_rewards:
             try:
-                distribute_static_rewards(
+                reward_writes.append(distribute_static_rewards(
                     driver=self.driver,
                     foundation_reward=self.static_rewards_amount_foundation,
                     master_reward=self.static_rewards_amount_validators,
-                )
+                ))
             except Exception as e:
                 print(f"REWARD ERROR: {e}, No reward distributed for this block")
 
         if self.current_block_rewards:
             for tx_hash, reward in self.current_block_rewards.items():
                 try:
-                    distribute_rewards(
+                    reward_writes.append(distribute_rewards(
                         stamp_rewards_amount=reward["amount"],
                         stamp_rewards_contract=reward["contract"],
                         driver=self.driver,
                         client=self.client,
-                    )
+                    ))
                 except Exception as e:
                     print(f"REWARD ERROR: {e}, No reward distributed for {tx_hash}")
+           
+        reward_hash = hash_from_rewards(reward_writes)
+        self.fingerprint_hashes.append(reward_hash)
+        fingerprint_hash = hash_list(self.fingerprint_hashes)
+
+        return ResponseFinalizeBlock(validator_updates=self.validator_handler.build_validator_updates(), tx_results=tx_results, app_hash=fingerprint_hash)
+
 
+    def commit(self) -> ResponseCommit:
+        fingerprint_hash = hash_list(self.fingerprint_hashes)
```

Comment:
any reason we're hashing the list twice in one block ? shouldn't it just be done once, after all the txs are processed ?